In [2]:
# Importar librerías
import pandas as pd

# 1. Analisis inicial de datos

In [3]:
# Cargamos el dataset desde el link
billboard = pd.read_csv('https://raw.githubusercontent.com/hadley/tidy-data/master/data/billboard.csv', warn_bad_lines=False, error_bad_lines=False, encoding='iso-8859-1')

In [4]:
# Observamos las dimensiones del dataset
billboard.shape

(317, 83)

In [5]:
# Vemos las columnas
billboard.columns

Index(['year', 'artist.inverted', 'track', 'time', 'genre', 'date.entered',
       'date.peaked', 'x1st.week', 'x2nd.week', 'x3rd.week', 'x4th.week',
       'x5th.week', 'x6th.week', 'x7th.week', 'x8th.week', 'x9th.week',
       'x10th.week', 'x11th.week', 'x12th.week', 'x13th.week', 'x14th.week',
       'x15th.week', 'x16th.week', 'x17th.week', 'x18th.week', 'x19th.week',
       'x20th.week', 'x21st.week', 'x22nd.week', 'x23rd.week', 'x24th.week',
       'x25th.week', 'x26th.week', 'x27th.week', 'x28th.week', 'x29th.week',
       'x30th.week', 'x31st.week', 'x32nd.week', 'x33rd.week', 'x34th.week',
       'x35th.week', 'x36th.week', 'x37th.week', 'x38th.week', 'x39th.week',
       'x40th.week', 'x41st.week', 'x42nd.week', 'x43rd.week', 'x44th.week',
       'x45th.week', 'x46th.week', 'x47th.week', 'x48th.week', 'x49th.week',
       'x50th.week', 'x51st.week', 'x52nd.week', 'x53rd.week', 'x54th.week',
       'x55th.week', 'x56th.week', 'x57th.week', 'x58th.week', 'x59th.week',
       '

In [6]:
# Observamos las primeras filas del dataset 
billboard.head()

year      artist.inverted  ... x75th.week x76th.week
0  2000      Destiny's Child  ...        NaN        NaN
1  2000              Santana  ...        NaN        NaN
2  2000        Savage Garden  ...        NaN        NaN
3  2000              Madonna  ...        NaN        NaN
4  2000  Aguilera, Christina  ...        NaN        NaN

[5 rows x 83 columns]

1. **¿Cuáles son las observaciones?**

  Son las canciones que estuvieron el top 100 de billboard 

2. **¿Cuáles son las variables?**

  Año, nombre del artista, canción, duración, género, fecha en la que entró al chart, fecha en la que llegó a la máxima posición, ranking que tuvo en la primera semana,  ranking que tuvo en la segunda semana, etc.



# 2. Preguntas específicas

Para contestar preguntas especificas: realizar filtros y selecciones de columnas existen dos tipos de métodos:

* **Método basado en indexing**: `DataFrame.loc[selección filas, selección columnas]`

* **Método basado en expresiones**: `DataFrame.query(expresión)`

Queremos hacer un análisis sobre estos datos con las siguientes preguntas:

a. ¿Cuáles géneros de música hay?

b. ¿Cuáles son las canciones de género Electronica? 

c. ¿Cuáles son los/las artistas y canciones del género Rock que ingresaron al ranking a partir de octubre? *italicized text*

### a. ¿Cuáles géneros de música hay?

Para contestar esta pregunta podemos usar el método `DataFrame.column.unique()` o la funcion `pd.unique()` que nos permiten obtener los valores únicos de una columna

In [7]:
# Vemos los generos distintos que existen
pd.unique(billboard.genre)

array(['Rock', 'Latin', 'Country', 'Rap', 'Pop', 'Electronica', 'Jazz',
       'R&B', 'Reggae', 'Gospel'], dtype=object)

In [8]:
# La alternativa
billboard.genre.unique()

array(['Rock', 'Latin', 'Country', 'Rap', 'Pop', 'Electronica', 'Jazz',
       'R&B', 'Reggae', 'Gospel'], dtype=object)

### b. ¿Cuáles son las canciones de género Electronica? 
Veamos los dos métodos para contestar esta pregunta

In [9]:
# Metodo basado en indexing
## Ponemos la condicion del filtro
billboard.loc[billboard.genre=='Electronica']

year artist.inverted  ... x75th.week x76th.week
43   2000       Eiffel 65  ...        NaN        NaN
109  2000    Alice Deejay  ...        NaN        NaN
265  2000     Fatboy Slim  ...        NaN        NaN
288  2000  Madison Avenue  ...        NaN        NaN

[4 rows x 83 columns]

In [10]:
# Para seleccionar solo la cancion lo añadimos en el segundo argumento del .loc[]
billboard.loc[billboard.genre=='Electronica', 'track']

43          Blue (Da Ba Dee)
109         Better Off Alone
265    The Rockafeller Skank
288       Don't Call Me Baby
Name: track, dtype: object

In [11]:
# Metodo basado en expresiones
## Definimos el filtro mediante la expresion
billboard.query("genre=='Electronica'")

year artist.inverted  ... x75th.week x76th.week
43   2000       Eiffel 65  ...        NaN        NaN
109  2000    Alice Deejay  ...        NaN        NaN
265  2000     Fatboy Slim  ...        NaN        NaN
288  2000  Madison Avenue  ...        NaN        NaN

[4 rows x 83 columns]

El **método basado en expresiones** puede ser más sencillo de leer pero no se puede seleccionar columnas directamente mientras que el **método basado en indexing** puede ser más complejo de leer pero permite seleccionar filas y columnas directamente

In [12]:
# Luego de utilizar query podemos seleccionar las columnas
billboard.query("genre=='Electronica'")['track']

43          Blue (Da Ba Dee)
109         Better Off Alone
265    The Rockafeller Skank
288       Don't Call Me Baby
Name: track, dtype: object

### c. ¿Cuáles son los/las artistas y canciones del género Rock que ingresaron al ranking a partir de octubre? 

Para contestar esta pregunta debemos filtrar por el genero y la fecha de ingreso y quedarnos con las columnas de artista y canción

Para indicar que se deben cumplir ambas condiciones utilizamos el simbolo `&` entre ambos filtros: (condicion 1) & (condicion 2). Esto significa que se deben cumplir ambas condiciones en simultaneo

In [13]:
# Metodo basado en indexing
billboard.loc[(billboard.genre=='Rock') & (billboard['date.entered'] >= '2000-10-01'), ['artist.inverted','track']]

artist.inverted               track
52   Backstreet Boys, The   Shape Of My Heart
102                 Avant       My First Love
167             Vitamin C            The Itch
170          Thomas, Carl           Emotional
193          3 Doors Down               Loser
211                 Usher       Pop Ya Collar
224           Limp Bizkit             Rollin'
232             Nine Days             If I Am
237        Offspring, The  Original Prankster
247      Wallflowers, The         Sleepwalker
279         Moore, Chante         Straight Up
308         Braxton, Toni      Spanish Guitar
309               Tuesday              I Know

# 3. Preguntas específicas II

Ya pudimos responder nuestra preguntas previas y ahora queremos hacer preguntas más complejas vinculadas a la posición en el ranking en distintas semanas, como por ejemplo:

* ¿En qué semanas estuvo en la posición 1 cada una de las canciones de Rock que ingresaron antes de la segunda mitad del año 2000?

¿Es posible contestar esta pregunta de manera sencilla?

No, parece ser algo bastante complicado.


# 4. Tidy data

Recordemos que las condiciones para que un dataset se considere ordenado (tidy) son:

* Cada variable forma una columna
* Cada observación forma una fila
* Cada tipo de unidad observacional forma una tabla

¿El dataset de billboard cumple con estas condiciones?

No. Observamos que:

1. La variable del número de semana en el ranking no está como una columna
2. Aunque es discutible, podríamos argumentar que hay dos unidades observaciones en la misma tabla: características de la canción y posición en el ranking

Para resolver el primer problema vamos a utilizar la función `pd.melt()`

In [14]:
# Al ejecutar esta celda se abre la documentación de la función
? pd.melt()

Nuestro primer paso va a ser especificar las **id_vars**, variables identificadoras y las **value_vars**.

¿Cuáles son cada una de ellas?

**id_vars**: "year", "artist.inverted", "track", "time", "genre", "date.entered", "date.peaked"

Aunque si tenemos en cuenta la pregunta puntual pueden ser menos

**value_vars**: todas las variables de las semanas

Tengamos en cuenta que no es necesario poner la lista de todas las variables ya que la función considera como value_vars a todas aquellas columnas que no aparecen en id_vars

In [15]:
# Aplicamos la función melt y tomamos una muestra
pd.melt(billboard,
        id_vars=["year", "artist.inverted", "track", "time", "genre", "date.entered", "date.peaked"]).sample(5)

year  artist.inverted              track  ... date.peaked    variable value
9985   2000    Tippin, Aaron          Kiss This  ...  2000-10-21  x32nd.week   NaN
11961  2000        Nine Days            If I Am  ...  2000-12-02  x38th.week   NaN
3493   2000  Destiny's Child        Say My Name  ...  2000-03-18  x12th.week  14.0
7223   2000      Limp Bizkit     N 2 Gether Now  ...  2000-01-08  x23rd.week   NaN
22345  2000     Walker, Clay  The Chain Of Love  ...  2000-06-10  x71st.week   NaN

[5 rows x 9 columns]

Vemos que ahora tenemos el número de la semana como una columna y el ranking en esa semana como otra.

Pero, podemos mejorar el resultado y que nuestras nuevas columnas no tengan nombres genericos como "variable" y "value". Esto lo hacemos utilizando los argumentos **var_name** y **value_name**

In [16]:
# Creamos nuestro dataset tidy
billboard_tidy = pd.melt(billboard, 
                      id_vars=["year", "artist.inverted", "track", "time", "genre", "date.entered", "date.peaked"], 
                      var_name="week", 
                      value_name="ranking")

In [18]:
billboard_tidy.head()

year      artist.inverted  ...       week ranking
0  2000      Destiny's Child  ...  x1st.week    78.0
1  2000              Santana  ...  x1st.week    15.0
2  2000        Savage Garden  ...  x1st.week    71.0
3  2000              Madonna  ...  x1st.week    41.0
4  2000  Aguilera, Christina  ...  x1st.week    57.0

[5 rows x 9 columns]

In [20]:
# Dimensiones del dataset
billboard_tidy.shape

(24092, 9)

Ahora el dataset tiene 24092 filas y 9 columnas en comparación a las 317 filas y 83 columnas originales.

Si bien podemos realizar algunas transformaciones adicionales para mejorar el dataset, son necesarias técnicas que veremos más adelante en el curso. Pero ya lo tenemos preparado para contestar la pregunta: ¿En qué semanas estuvo en la posición 1 cada una de las canciones de Rock que ingresaron antes de la segunda mitad del año 2000?

In [30]:
resultado = billboard_tidy.loc[(billboard_tidy['genre']=='Rock')&(billboard_tidy['date.entered']<='2000-06-30')&(billboard_tidy['ranking']==1), ['artist.inverted', 'track', 'week']]
resultado

artist.inverted                  track        week
2227                Sisqo             Incomplete   x8th.week
2230  Aguilera, Christina      What A Girl Wants   x8th.week
2537              Santana           Maria, Maria   x9th.week
2544                Sisqo             Incomplete   x9th.week
2547  Aguilera, Christina      What A Girl Wants   x9th.week
2854              Santana           Maria, Maria  x10th.week
3171              Santana           Maria, Maria  x11th.week
3175                Janet  Doesn't Really Matter  x11th.week
3186        Carey, Mariah  Thank God I Found You  x11th.week
3488              Santana           Maria, Maria  x12th.week
3492                Janet  Doesn't Really Matter  x12th.week
3805              Santana           Maria, Maria  x13th.week
3809                Janet  Doesn't Really Matter  x13th.week
3810      Destiny's Child            Say My Name  x13th.week
3814               N'Sync       It's Gonna Be Me  x13th.week
3819      matchbox twenty                   Bent  x13th.week
4122              Santana           Maria, Maria  x14th.week
4127      Destiny's Child            Say My Name  x14th.week
4131               N'Sync       It's Gonna Be Me  x14th.week
4135              Aaliyah              Try Again  x14th.week
4439              Santana           Maria, Maria  x15th.week
4440        Savage Garden     I Knew I Loved You  x15th.week
4444      Destiny's Child            Say My Name  x15th.week
4756              Santana           Maria, Maria  x16th.week
4757        Savage Garden     I Knew I Loved You  x16th.week
5073              Santana           Maria, Maria  x17th.week
5074        Savage Garden     I Knew I Loved You  x17th.week
5390              Santana           Maria, Maria  x18th.week
5708        Savage Garden     I Knew I Loved You  x19th.week
7937     Vertical Horizon    Everything You Want  x26th.week
8255                Creed    With Arms Wide Open  x27th.week

In [31]:
# Podemos mejorar el resultado ordenando el resultado con la función sort_values
resultado.sort_values(['track', 'week'])

artist.inverted                  track        week
3819      matchbox twenty                   Bent  x13th.week
3175                Janet  Doesn't Really Matter  x11th.week
3492                Janet  Doesn't Really Matter  x12th.week
3809                Janet  Doesn't Really Matter  x13th.week
7937     Vertical Horizon    Everything You Want  x26th.week
4440        Savage Garden     I Knew I Loved You  x15th.week
4757        Savage Garden     I Knew I Loved You  x16th.week
5074        Savage Garden     I Knew I Loved You  x17th.week
5708        Savage Garden     I Knew I Loved You  x19th.week
2227                Sisqo             Incomplete   x8th.week
2544                Sisqo             Incomplete   x9th.week
3814               N'Sync       It's Gonna Be Me  x13th.week
4131               N'Sync       It's Gonna Be Me  x14th.week
2854              Santana           Maria, Maria  x10th.week
3171              Santana           Maria, Maria  x11th.week
3488              Santana           Maria, Maria  x12th.week
3805              Santana           Maria, Maria  x13th.week
4122              Santana           Maria, Maria  x14th.week
4439              Santana           Maria, Maria  x15th.week
4756              Santana           Maria, Maria  x16th.week
5073              Santana           Maria, Maria  x17th.week
5390              Santana           Maria, Maria  x18th.week
2537              Santana           Maria, Maria   x9th.week
3810      Destiny's Child            Say My Name  x13th.week
4127      Destiny's Child            Say My Name  x14th.week
4444      Destiny's Child            Say My Name  x15th.week
3186        Carey, Mariah  Thank God I Found You  x11th.week
4135              Aaliyah              Try Again  x14th.week
2230  Aguilera, Christina      What A Girl Wants   x8th.week
2547  Aguilera, Christina      What A Girl Wants   x9th.week
8255                Creed    With Arms Wide Open  x27th.week